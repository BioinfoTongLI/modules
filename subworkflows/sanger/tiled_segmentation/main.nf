include { BIOINFOTONGLI_CELLPOSE as CELLPOSE } from '../../../modules/sanger/bioinfotongli/cellpose/main'
include { BIOINFOTONGLI_GENERATETILES as GENERATE_TILE_COORDS } from '../../../modules/sanger/bioinfotongli/generatetiles/main'

params.images = [
    [["id":"test1"], "file1"],
    [["id":"test2"], "file2"],
]
params.cell_diameters = [30, 40]
params.cellpose_model_dir = "/lustre/scratch126/cellgen/cellgeni/tl10/cellpose_models"
params.debug=false

container_version = "0.1.0"


process MERGE_OUTLINES {
    tag "${meta.id}"
    debug params.debug

    label "medium_mem"

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "quay.io/bioinfotongli/tiled_cellpose:${container_version}":
        "quay.io/bioinfotongli/tiled_cellpose:${container_version}"}"

    publishDir params.out_dir + "/cellpose_segmentation_merged_wkt"

    input:
    tuple val(meta), val(cell_diameter), path(wkts)

    output:
    tuple val(meta), val(cell_diameter), path("${stem}_merged.wkt"), emit: merged_wkt
    tuple val(meta), val(cell_diameter), path("${stem}_merged.geojson"), emit: merged_geojson
    path "versions.yml"           , emit: versions

    script:
    stem = "${meta.id}_diam-${cell_diameter}"
    def args = task.ext.args ?: ''  
    """
    /opt/conda/bin/python /scripts/merge_wkts.py run \
        --sample_id "${stem}" \
        ${wkts} \
        ${args}
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(/scripts/merge_wkts.py version 2>&1) | sed 's/^.*merge_wkts.py //; s/Using.*\$//' ))
    END_VERSIONS
    """
}


workflow TILED_SEGMENTATION {
    take:
    images 
    method

    main:
    ch_versions = Channel.empty()
    GENERATE_TILE_COORDS(images)
    ch_versions = ch_versions.mix(GENERATE_TILE_COORDS.out.versions.first())

    images_tiles = GENERATE_TILE_COORDS.out.tile_coords.splitCsv(header:true, sep:",").map{ meta, coords ->
        [meta, coords.X_MIN, coords.Y_MIN, coords.X_MAX, coords.Y_MAX]
    }

    if (method == "CELLPOSE") {
        CELLPOSE(images_tiles.combine(images, by:0).combine(channel.from(params.cell_diameters)))
        wkts = CELLPOSE.out.wkts.groupTuple(by:[0,1])
        ch_versions = ch_versions.mix(CELLPOSE.out.versions.first())
    } else if (method == "STARDIST") {
        STARDIST(images_tiles)
        wkts = STARDIST.out.wkts.groupTuple(by:[0,1])
        ch_versions = ch_versions.mix(STARDIST.out.versions.first())
    } else {
        error "Invalid segmentation method: ${method}"
    }
    MERGE_OUTLINES(wkts)
    ch_versions = ch_versions.mix(MERGE_OUTLINES.out.versions.first())

    emit:
    wkt         = MERGE_OUTLINES.out.merged_wkt   // channel: [ val(meta), [ wkt ] ]
    versions    = ch_versions                     // channel: [ versions.yml ]
}