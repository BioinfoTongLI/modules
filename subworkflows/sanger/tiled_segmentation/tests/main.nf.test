nextflow_workflow {

    name "Test Subworkflow TILED_SEGMENTATION"
    script "../main.nf"
    config "./nextflow.config"
    workflow "TILED_SEGMENTATION"

    tag "subworkflows"
    tag "subworkflows_sanger"
    tag "subworkflows/tiled_segmentation"

    test("mouse_heart - tif") {

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    file(params.modules_testdata_base_path + 'imaging/segmentation/nuclear_image.tif', checkIfExists: true),
                ]
                input[1] = 'CELLPOSE'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mouse_heart - tif - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test'], // meta map
                    file(params.modules_testdata_base_path + 'imaging/segmentation/nuclear_image.tif', checkIfExists: true),
                ]
                input[1] = 'CELLPOSE'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}
