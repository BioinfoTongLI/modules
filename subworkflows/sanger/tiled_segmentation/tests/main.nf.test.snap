{
    "nuclear - tif - instanseg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,9fe55ab05e13024faee6daa2ddcf8e6d"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ],
                "wkt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,9fe55ab05e13024faee6daa2ddcf8e6d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-18T22:25:28.496236908"
    },
    "nuclear - tif - cellpose": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,0d1072acebefc19c48bbcdb17ac57d5f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,7e24ffbd0425ebf6bf537e55ac367001"
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,7e24ffbd0425ebf6bf537e55ac367001"
                ],
                "wkt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,0d1072acebefc19c48bbcdb17ac57d5f"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-18T22:24:29.505111421"
    },
    "nuclear - tif - deepcell": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,a3f7237d662627863a30cff6128b989c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ],
                "wkt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,a3f7237d662627863a30cff6128b989c"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-18T22:26:52.113552786"
    },
    "nuclear - tif - stardist": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e3945a4be7e4f9daa0ab9c717a9cce10"
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e3945a4be7e4f9daa0ab9c717a9cce10"
                ],
                "wkt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-18T22:24:58.557258383"
    }
}